### PR TITLE
[Merged by Bors] - feat: minimal port of Fintype

### PR DESCRIPTION
**Pull request by kim-em**
This is a bare minimum port of the definitions of `Finset` / `Multiset` / `Fintype`. I am hoping that these are enough to port the `fin_cases` tactic. (i.e. replacing the unfinished work in #346, which still had sorries in the corresponding material)

Where possible I have started by copying and pasting the output from `mathlib3port`.

### CODE REVIEW

**Comment by digama0** [comment]
Note: the commit convention for lean 4 no longer puts the file name in the commit message.

**Comment by semorrison** [comment]
@digama0, the linter is [asking](https://github.com/leanprover-community/mathlib4/actions/runs/3086289246/jobs/4990491062) for doc-strings on the fields on structures. I wouldn't totally object to switching to this standard, but I'm pretty sure this is not intentional. If you agree, I can investigate fixing the linter.

**Comment by digama0** [comment]
It is intentional, although I don't mind nolinting the definitions if you think we should delay documenting it.

File: Mathlib/Data/Finset/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+universe u
+
+variable {α : Type _} {β : Type _} {γ : Type _}
+
+/-- `finset α` is the type of finite sets of elements of `α`. It is implemented
+  as a multiset (a list up to permutation) which has no duplicate elements. -/
+structure Finset (α : Type _) where
+  val : Multiset α
+  Nodup : Nodup val
```
Comment (digama0) [inline Mathlib/Data/Finset/Basic.lean:49]: This should be lowercase because it's a proof. Types / Props are uppercase, proofs / functions are lowercase.

**Comment by digama0** [comment]
I made some minor style changes and golfed the proof of `perm_inv_core`. If it all looks good to you then:

bors d+

**Comment by semorrison** [comment]
Thanks, looks great.

bors merge

**Pull request closed without merging**